source/uMod_DX9_dll.cpp

```diff
@@ -191,133 +182,32 @@ IDirect3D9* APIENTRY uMod_Direct3DCreate9(UINT SDKVersion)
+    ASSERT(pIDirect3D9_orig);
 
-    // we detour again
-    Direct3DCreate9_fn = static_cast<Direct3DCreate9_type>(DetourFunc((BYTE*)Direct3DCreate9_fn, (BYTE*)uMod_Direct3DCreate9, 5));
-    /*
-    if (Direct3DCreate9Ex_fn!=NULL)
-    {
-      Direct3DCreate9Ex_fn = (Direct3DCreate9Ex_type)DetourFunc( (BYTE*) Direct3DCreate9Ex_fn, (BYTE*)uMod_Direct3DCreate9Ex,7);
-    }
-  */
-    return pIDirect3D9; //return our object instead of the "real one"
+    return new uMod_IDirect3D9(pIDirect3D9_orig); //return our object instead of the "real one"
 }
 
 HRESULT APIENTRY uMod_Direct3DCreate9Ex(UINT SDKVersion, IDirect3D9Ex** ppD3D)
 {
     Message("uMod_Direct3DCreate9Ex:  original %p, uMod %p\n", Direct3DCreate9Ex_fn, uMod_Direct3DCreate9Ex);
 
-    // in the Internet are many tutorials for detouring functions and all of them will work without the following 5 marked lines
-    // but somehow, for me it only works, if I retour the function and calling afterward the original function
-
-    // BEGIN
-
-    LoadOriginalDll();
-    /*
-    if (Direct3DCreate9_fn!=NULL)
-    {
-      RetourFunc((BYTE*) GetProcAddress( gl_hOriginalDll, "Direct3DCreate9"), (BYTE*)Direct3DCreate9_fn, 5);
-      Direct3DCreate9_fn = (Direct3DCreate9_type) GetProcAddress( gl_hOriginalDll, "Direct3DCreate9");
-    }
-  */
-    RetourFunc((BYTE*)GetProcAddress(gl_hOriginalDll, "Direct3DCreate9Ex"), (BYTE*)Direct3DCreate9Ex_fn, 7);
-    Direct3DCreate9Ex_fn = (Direct3DCreate9Ex_type)GetProcAddress(gl_hOriginalDll, "Direct3DCreate9Ex");
-    // END
+    ASSERT(Direct3DCreate9Ex_ret);
 
     IDirect3D9Ex* pIDirect3D9Ex_orig = nullptr;
-    HRESULT ret;
-    if (Direct3DCreate9Ex_fn) {
-        ret = Direct3DCreate9Ex_fn(SDKVersion, &pIDirect3D9Ex_orig); //creating the original IDirect3D9 object
-    }
-    else {
-        return D3DERR_NOTAVAILABLE;
-    }
+    HRESULT ret = Direct3DCreate9Ex_fn(SDKVersion, &pIDirect3D9Ex_orig); //creating the original IDirect3D9 object
 
-    if (ret != S_OK) {
+    if (ret != S_OK)
         return ret;
-    }
 
-    uMod_IDirect3D9Ex* pIDirect3D9Ex;
-    if (pIDirect3D9Ex_orig) {
-        pIDirect3D9Ex = new uMod_IDirect3D9Ex(pIDirect3D9Ex_orig); //creating our uMod_IDirect3D9 object
-    }
-
-    // we detour again
-    /*
-      if (Direct3DCreate9_fn!=NULL)
-      {
-        Direct3DCreate9_fn = (Direct3DCreate9_type)DetourFunc( (BYTE*) Direct3DCreate9_fn, (BYTE*)uMod_Direct3DCreate9,5);
-      }
-      */
-    Direct3DCreate9Ex_fn = static_cast<Direct3DCreate9Ex_type>(DetourFunc((BYTE*)Direct3DCreate9Ex_fn, (BYTE*)uMod_Direct3DCreate9Ex, 7));
-    ppD3D = (IDirect3D9Ex**)&pIDirect3D9Ex; //return our object instead of the "real one"
+    // @Cleanup: should be we freeing pIDirect3D9Ex at the end of our own lifecycle?
+    uMod_IDirect3D9Ex* pIDirect3D9Ex = new uMod_IDirect3D9Ex(pIDirect3D9Ex_orig);
+    ppD3D = (IDirect3D9Ex**)&pIDirect3D9Ex;
```

Comment:
wouldn't this be a dangling reference? nothing is actually written here, but if it were, it would certainly lead to problems

Reply:
Which is why I added the comment, yes - the current version of uMod also does this, but technically theres not a good time to know when to free it until the program itself closes